shortfin/python/shortfin_apps/sd/components/generate.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2024 Advanced Micro Devices, Inc.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import asyncio
+import io
+import logging
+
+import shortfin as sf
+import shortfin.array as sfnp
+
+# TODO: Have a generic "Responder" interface vs just the concrete impl.
+from shortfin.interop.fastapi import FastAPIResponder
+
+from .io_struct import GenerateReqInput
+from .messages import InferenceExecRequest, InferencePhase
+from .service import GenerateService
+
+logger = logging.getLogger(__name__)
+
+
+class GenerateImageProcess(sf.Process):
+    """Process instantiated for every image generation.
+
+    This process breaks the sequence into individual inference and sampling
+    steps, submitting them to the batcher and marshaling final
+    results.
+
+    Responsible for a single image.
+    """
+
+    def __init__(
+        self,
+        client: "ClientGenerateBatchProcess",
+        gen_req: GenerateReqInput,
+        index: int,
+    ):
+        super().__init__(fiber=client.fiber)
+        self.client = client
+        self.gen_req = gen_req
+        self.index = index
+        self.result_image = None
+
+    async def run(self):
+        # TODO: make sure gen_req is being indexed for the singular image generation sequence's inputs.
+        exec = InferenceExecRequest(
+            InferencePhase.PREPARE,
+            self.gen_req.prompt[self.index],
+            self.gen_req.neg_prompt[self.index],
+            self.gen_req.height[self.index],
+            self.gen_req.width[self.index],
+            self.gen_req.steps[self.index],
+            self.gen_req.guidance_scale[self.index],
+            self.gen_req.seed[self.index],
+        )
+        self.client.batcher.submit(exec)
+        await exec.done
+
+        exec.reset(InferencePhase.ENCODE)
+        self.client.batcher.submit(exec)
+        await exec.done
```

Comment:
I still think we have this broken up into way too many phases at the high level here. It will block a lot of pipelining to do it this way. It would be better to add the needed phases to one request object and submit it at once: then have the executor perform each step in sequence vs batching/waiting for each step on the outside like this.

Fine to submit as-is but something we will want to invert.

Reply:
I agree! This is much clearer now that it’s laid out and loading/batching enhancements become more tangible.

I will follow up with a separate patch